addons/account_edi/models/account_edi_format.py

```diff
@@ -506,21 +506,90 @@ def _retrieve_partner(self, name=None, phone=None, mail=None, vat=None):
+                normalized_vat = normalized_vat[2:]
+                partner = self.env['res.partner'].search(extra_domain + [
+                    ('vat', '=', normalized_vat),
+                    ('country_id.code', '=', country_prefix.lower()),
+                ], limit=1)
+
+                # The country could be not specified.
+                if not partner:
+                    partner = self.env['res.partner'].search(extra_domain + [
+                        ('vat', '=', normalized_vat),
+                        ('country_id', '=', False),
+                    ], limit=1)
+
+            # The vat could be a string of alphanumeric values without country code but with missing zeros at the
+            # beginning.
+            if not partner:
+                try:
+                    vat_only_alphanumeric = str(int(normalized_vat))
+                except ValueError:
+                    vat_only_alphanumeric = None
+
+                if vat_only_alphanumeric:
+                    query = self.env['res.partner']._where_calc([('active', '=', True)], extra_domain)
+                    tables, where_clause, where_params = query.get_sql()
+
+                    self._cr.execute(f'''
+                        SELECT res_partner.id
+                        FROM {tables}
+                        WHERE {where_clause}
+                        AND res_partner.vat ~ %s
+                        LIMIT 1
+                    ''', where_params + ['^0*%s$' % vat_only_alphanumeric])
+                    partner_row = self._cr.fetchone()
+                    if partner_row:
+                        partner = self.env['res.partner'].browse(partner_row[0])
+
+            return partner
+
+        def search_with_phone_mail(extra_domain):
+            domains = []
+            if phone:
+                domains.append([('phone', '=', phone)])
+                domains.append([('mobile', '=', phone)])
+            if mail:
+                domains.append([('email', '=', mail)])
+
+            if not domains:
+                return None
+
+            domain = expression.OR(domains)
+            if extra_domain:
+                domain = expression.AND([domain, extra_domain])
+            return self.env['res.partner'].search(domain, limit=1)
+
+        def search_with_name(extra_domain):
+            if not name:
+                return None
+            return self.env['res.partner'].search([('name', 'ilike', name)] + extra_domain, limit=1)
+
+        for extra_domain, search_method in [
+            ([('company_id', '=', self.env.company.id)], search_with_vat),
+            ([('company_id', '=', False)], search_with_vat),
+            ([('company_id', '=', self.env.company.id)], search_with_phone_mail),
+            ([('company_id', '=', False)], search_with_phone_mail),
+            ([('company_id', '=', self.env.company.id)], search_with_name),
+            ([('company_id', '=', False)], search_with_name),
+        ]:
```

Comment:
Make two loops: one on the functions and one on the extra domain, instead of creating all the domains by hand
It would also make sense to try all the functions with the domain restricting on the company then all the functions without restricting on the company